### DISCUSSION TITLE
Android 11+ local HTML using StageWebView

### DISCUSSION THREAD

**Discussion by KarlSchumann** [airsdk/Adobe-Runtime-Support#1365] [Q&A] [answered]
Is there anyway to load local HTML files that are in the applicationStorageDirectory in a StageWebView when targeting Android 11+?

Here is an example of code that has worked previously and is still working for iOS:
```
_webView = new StageWebView();
_webView.stage = this.stage; 

var appStorage:File = File.applicationStorageDirectory;
appStorage = appStorage.resolvePath("help")
appStorage = appStorage.resolvePath("help.html");
var fPath:String = appStorage.nativePath;
var file:File = new File(fPath);
fPath = file.url;
_webView.loadURL(fPath);
```

This is now resulting in the following:
![image](https://user-images.githubusercontent.com/26549853/142064663-08e03b27-b133-47e6-82fc-7ac5a0375d07.png)

While digging in to this issue I have tried adding the following to the application manifest, which doesn't resolve it as it seems to be more for dealing with http vs https:
`android:usesCleartextTraffic="true"`

It seems like the solution when coding native would be to do the following, but doesn't seem like that is available through AIR?
`webView.settings.allowFileAccess = true`

Anyone know of any workaround or better approach for handling this?

**Comment by ajwfrost** [comment, marked as answer]
Hi - @KarlSchumann  you're right in terms of the `allowFileAccess` API:
https://developer.android.com/reference/android/webkit/WebSettings#setAllowFileAccess(boolean)

The default for this has changed from 'true' to 'false' and there's no way (currently) for AIR apps to flag this as true again.

Before we add this mechanism though, can I check why it's not possible to have the files available as part of your normal application bundle or as an asset pack? As these should still be accessible via the "file:///android_asset/..." mechanism. Android are trying to stop people from enabling uncontrolled/arbitrary content (from your app-storage folder) from being accessed within the webview control.. so although we may be able to update things to allow you you to set this value, I would expect that this capability may well be dropped in a future Android version...

thanks

**Reply by KarlSchumann** [reply]
@ajwfrost Thanks for the suggestion I will give that a shot! I thought files in the applicationStorageDirectory and the applicationDirectory would have the same restrictions. As to why it was necessary for them to be in storage in the first place there doesn't seem to be a great reason the only indication I have was it being necessary was from a comment in our codebase from ~2015

`// Copy files to the applicationStorageDirectory (localstore) that cannot be read from the applicationDirectory using stageWebView`

So seems like maybe at some point we were having issues loading them up as part of the application directory and moving them to storage worked for us and stayed that way. This app is published for iOS and Android so not sure if maybe one needed it one way at that time.

**Reply by ajwfrost** [reply]
Ah it may be something that was done in order to make it work the same on both iOS and Android then .. plus of course with Android you wouldn't be able to use `File.applicationDirectory.resolve('filename').url` (or .nativePath) because Android needs you to call it `file:///android_assets/filename`. So that means different code for Android and for iOS...

One of the other folks who contacted us about this has got some files that the application generates that will need to be viewed from the application storage folder, so we are actually going to add a flag that can be set up to permit this i.e. to tell Android to allow file:// URLs. So, if you can wait a couple of weeks, that might be a simpler option and would mean you don't need to change any code...

thanks

**Reply by balaedit** [reply]
Is there any update in opening/viewing local pdf (applicationStorageDirectory) in android using any syntax such as stagewebview?

**Reply by ajwfrost** [reply]
We're working on an ANE to get this sort of functionality working. The general problem is that although you may own the PDF, it needs additional steps now to provide the right permissions for the PDF viewer to display it.

I guess it should be possible to have the StageWebView display this though? you'd need to have the application descriptor to enable file access for the Android webview:
```<android><webViewAllowFileAccess>true</webViewAllowFileAccess>...</android>``` 

thanks

**Reply by balaedit** [reply]
I added the given line (`<webViewAllowFileAccess>true</webViewAllowFileAccess>`) in the manifest.

But noting is occurred.

my package is given here : 
[html.zip](https://github.com/airsdk/Adobe-Runtime-Support/files/9351725/html.zip)

Please suggest and given some solution to open pdf.

**Reply by teotigraphix** [reply]
> One of the other folks who contacted us about this has got some files that the application generates that will need to be viewed from the application storage folder, so we are actually going to add a flag that can be set up to permit this i.e. to tell Android to allow file:// URLs. So, if you can wait a couple of weeks, that might be a simpler option and would mean you don't need to change any code...


I am running into this as well @ajwfrost . I have the audio engine in **NDK** land, so it's trying to load files from the `applicationDirectory` which it could do before Android 11.

For now I am doing as said above, copying the synth preset files from `applicationDirectory` to the `applicationStorageDirectory` because I get a real `nativePath` to send to the **NDK** level.

Could you do a slight write up on this Android/iOS file stuff when AIR gets it worked out? This is still confusing what you are talking about with `file:///android_assets/filename` etc accessing what?

Thanks!
